tests/test_browser.go

```diff
@@ -281,62 +281,75 @@ func (b *testBrowser) awaitWithTimeout(timeout time.Duration, fn func() error) e
-	promise *goja.Promise, onFulfilled interface{}, onRejected ...interface{},
-) testPromise {
+func (b *testBrowser) promiseAll(promises ...*goja.Promise) testPromise {
 	b.t.Helper()
 	rt := b.runtime()
-	val, err := rt.RunString(
-		`(function(promise, onFulfilled, onRejected) { return promise.then(onFulfilled, onRejected) })`)
+	val, err := rt.RunString(`(function(...promises) { return Promise.all(...promises) })`)
 	require.NoError(b.t, err)
 	cal, ok := goja.AssertFunction(val)
 	require.True(b.t, ok)
-
-	switch len(onRejected) {
-	case 0:
-		val, err = cal(goja.Undefined(), rt.ToValue(promise), rt.ToValue(onFulfilled))
-	case 1:
-		val, err = cal(goja.Undefined(), rt.ToValue(promise), rt.ToValue(onFulfilled), rt.ToValue(onRejected[0]))
-	default:
-		panic("too many arguments to promiseThen")
+	valPromises := make([]goja.Value, len(promises))
+	for i, promise := range promises {
+		valPromises[i] = rt.ToValue(promise)
 	}
+	val, err = cal(goja.Undefined(), rt.ToValue(valPromises))
 	require.NoError(b.t, err)
 	newPromise, ok := val.Export().(*goja.Promise)
 	require.True(b.t, ok)
 
 	return testPromise{
 		Promise: newPromise,
+		call:    cal,
 		tb:      b,
 	}
 }
 
-type testPromise struct {
-	*goja.Promise
-	tb *testBrowser
-}
-
-//nolint:unparam // the result will be used in future tests
-func (t testPromise) then(onFulfilled interface{}, onRejected ...interface{}) testPromise {
-	return t.tb.promiseThen(t.Promise, onFulfilled, onRejected...)
-}
-
-func (b *testBrowser) promiseAll(promises ...*goja.Promise) *goja.Promise {
+func (b *testBrowser) promise(promise *goja.Promise) testPromise {
 	b.t.Helper()
 	rt := b.runtime()
-	val, err := rt.RunString(`(function(...promises) { return Promise.all(...promises) })`)
+	val, err := rt.RunString(`
+		(function(promise, resolve, reject) {
+			return promise.then(resolve, reject)
+		})
+	`)
 	require.NoError(b.t, err)
 	cal, ok := goja.AssertFunction(val)
 	require.True(b.t, ok)
-	valPromises := make([]goja.Value, len(promises))
-	for i, promise := range promises {
-		valPromises[i] = rt.ToValue(promise)
+
+	return testPromise{
+		Promise: promise,
+		call:    cal,
+		tb:      b,
 	}
-	val, err = cal(goja.Undefined(), rt.ToValue(valPromises))
-	require.NoError(b.t, err)
-	newPromise, ok := val.Export().(*goja.Promise)
-	require.True(b.t, ok)
+}
+
+type testPromise struct {
+	*goja.Promise
+	tb   *testBrowser
+	call goja.Callable
+}
+
+func (t testPromise) then(resolve interface{}, reject ...interface{}) testPromise {
+	var (
+		val goja.Value
+		err error
+		rt  = t.tb.runtime()
+	)
+	switch len(reject) {
+	case 0:
+		val, err = t.call(goja.Undefined(), rt.ToValue(t.Promise), rt.ToValue(resolve))
+	case 1:
+		val, err = t.call(goja.Undefined(), rt.ToValue(t.Promise), rt.ToValue(resolve), rt.ToValue(reject[0]))
+	default:
+		panic("too many arguments to promiseThen")
+	}
+	require.NoError(t.tb.t, err)
+
+	var ok bool
+	t.Promise, ok = val.Export().(*goja.Promise)
```

Comment:
This changes the original testPromise, IMO this shouldn't be the case you should make a new `testPromise`.

Reply:
That likely can reuse the `call` I think :thinking:

Reply:
Yeah, since there is a single promise function created in `promise`, it reuses it. Only changes the current `testPromise.Promise` so that the subsequent calls can use the new promise. Do you see any problem with that?

Reply:
Or, something like this:

```go
// then(...) {
//      ...
	p, ok := val.Export().(*goja.Promise)
	require.True(t.tb.t, ok)

	return t.tb.promise(p)
```

Reply:
Thanks, I've updated the test browser as we discussed.